heat_pump_adoption_modelling/pipeline/MCS/mcs_epc_joining.py

```diff
@@ -0,0 +1,322 @@
+
+    dhps["standardised_postcode"] = [
+        pc.upper().strip() for pc in dhps["postcode"].fillna("unknown")
+    ]
+
+    dhps["standardised_address"] = [
+        rm_punct(add1).lower().strip() + " " + rm_punct(add2).lower().strip()
+        for add1, add2 in zip(dhps.address_1.fillna(""), dhps.address_2.fillna(""))
+    ]
+
+    dhps["numeric_tokens"] = extract_numeric_tokens(dhps)
+
+    return dhps
+
+
+def prepare_epcs(epcs):
+    """Prepares Dataframe of EPC records by adding
+    standardised_postcode, standardised_address and numeric_tokens fields.
+
+    Parameters
+    ----------
+    epcs : pandas.Dataframe
+        Dataframe with POSTCODE and ADDRESS1 fields.
+
+    Return
+    ----------
+    epcs : pandas.Dataframe
+        Dataframe containing EPC records with added fields."""
+
+    # Remove spaces, uppercase and strip whitespace from
+    # postcodes in order to exact match on this field
+    epcs["standardised_postcode"] = [
+        postcode.replace(" ", "") for postcode in epcs["POSTCODE"].fillna("UNKNOWN")
+    ]
+
+    # Remove punctuation, lowercase and concatenate address fields
+    # for approximate matching
+    epcs["standardised_address"] = [
+        rm_punct(address).lower().strip() for address in epcs["ADDRESS1"]
+    ]
+
+    epcs["numeric_tokens"] = extract_numeric_tokens(epcs)
+
+    return epcs
+
+
+# ---------------------------------------------------------------------------------
+
+
+#### JOINING
+
+
+def form_matching(df1, df2):
+    """Forms a matching between two Dataframes.
+    Initially an index is formed between records with shared
+    standardised_postcode, then the records are compared for
+    exact matches on numeric tokens and fuzzy matches on
+    standardised_address (using Jaro-Winkler method).
+
+    Parameters
+    ----------
+    df1 : pandas.Dataframe
+        Dataframe with standardised_postcode, numeric_tokens
+        and standardised_address fields.
+    df2 : pandas.Dataframe
+        Dataframe with standardised_postcode, numeric_tokens
+        and standardised_address fields.
+
+    Return
+    ----------
+    matching : pandas.Dataframe
+        Dataframe giving indices of df1 and matched indices in df2
+        along with similarity scores for numeric tokens (0/1) and address (0-1)."""
+
+    # Index
+    print("- Forming an index...")
+    indexer = rl.Index()
+    indexer.block(on="standardised_postcode")
+    candidate_pairs = indexer.index(df1, df2)
+
+    # Compare
+    print("- Forming a comparison...")
+    comp = rl.Compare()
+    comp.exact("numeric_tokens", "numeric_tokens", label="numerics")
+    # Jaro-Winkler chosen here as it prioritises characters
+    # near the start of the string - this feels suitable as
+    # the key information (such as house name) is likely to
+    # be near the start of the string
+    # This feels better suited than e.g. Levenshtein as to not
+    # punish extra information at the end of the address field
+    # e.g. town, county
```

Comment:
This is so cool and essential it deserves to go into the doc string of the function above!

Reply:
Moved this